charms/kfp-metadata-writer/src/charm.py

```diff
@@ -46,8 +45,10 @@ def __init__(self, *args):
         )
 
         self.grpc_relation = self.charm_reconciler.add(
-            component=SdiRelationDataReceiverComponent(
-                charm=self, name="relation:grpc", relation_name="grpc"
+            component=K8sServiceInfoComponent(
+                charm=self,
+                relation_name=GRPC_RELATION_NAME,
+                refresh_event=self.on[GRPC_RELATION_NAME].relation_changed,
```

Comment:
Not a hard requirement, but I'd include the relation events in the `K8sServiceInfoComponent` class directly like [the sdi component does](https://github.com/canonical/charmed-kubeflow-chisme/blob/8a7454fede8dbcefe65cd45f2fc04a3926b7e421/src/charmed_kubeflow_chisme/components/serialised_data_interface_components.py#L84).  Main reason is I don't think there'd be a use case for us to need this component and not observe the relation events

Reply:
Oh sorry no, this is a bug.

The `KubernetesServiceInfoRequirer` already [observes the relation events](https://github1s.com/canonical/kfp-operators/blob/KF-5485-use-k8s-service-info-lib/charms/kfp-metadata-writer/lib/charms/mlops_libs/v0/k8s_service_info.py#L192-L200) - we don't need to observe those events here.  But, the `K8sServiceInfoComponent` **does need to observe the library's [`KubernetesServiceInfoUpdatedEvent`](https://github1s.com/canonical/kfp-operators/blob/KF-5485-use-k8s-service-info-lib/charms/kfp-metadata-writer/lib/charms/mlops_libs/v0/k8s_service_info.py#L143-L144) so the charm knows to wake whenever the lib sees updated data.  

Right now on a relation-data-changed event the underlying lib is handling the event and emitting a `KubernetesServiceInfoUpdatedEvent`, but nothing in the charm observes that so the charm itself is not doing anything with the new relation data.  You can see this by watching the `juju debug-log -i kfp-metadata-writer` while relating to mlmd - nothing happens and the charm stays blocked.  The charm does eventually recover from this Blocked state though when the next `update_status` event fires (triggering a full charm reconcile).

Reply:
This bug can be caught by a unit test, so we must have a test coverage gap.

I think the issue is that:
* [this test](https://github.com/canonical/kfp-operators/pull/436/files#diff-8e7738d6ff50a7d31913bee76270ebc0c80efefbcf3c75164fb3521f7bb4603eR62) asserts the charm *can access* the relation data, not that it actually used it to reconcile the charm after the relation was established
* [this test](https://github.com/canonical/kfp-operators/pull/436/files#diff-8e7738d6ff50a7d31913bee76270ebc0c80efefbcf3c75164fb3521f7bb4603eR132) does show we successfully use the data, but it does not prove the relation-changed event reconciled the charm either because it emits an `install` event right after the data is updated (the install event then **uses** the data, masking how the relation-changed event did not use the data)

If we move the `harness.charm.on.install.emit()` in the second test above the `harness.add_relation`, we now see the test fails

Reply:
Test it yourself to be sure, but I think [this pr against your pr](https://github.com/canonical/kfp-operators/pull/437) will fix this issue

Reply:
> but nothing in the charm observes that so the charm itself is not doing anything with the new relation data. You can see this by watching the juju debug-log -i kfp-metadata-writer while relating to mlmd - nothing happens and the charm stays blocked. The charm does eventually recover from this Blocked state though when the next update_status event fires (triggering a full charm reconcile).

yes, I observed exactly that behaviour when deploying the charm. I thought it was a limitation of the charmReconciler, though, as the library is in fact observing that exact event, and I kind of had to "force" it.

Maybe it's because I am less charmReconciler savvy, but it was not very obvious to me how to make this component observe non-standard events. I see [this guide](https://github.com/canonical/charmed-kubeflow-chisme/blob/main/src/charmed_kubeflow_chisme/components/how-to/authoring-a-simple-component.md#handling-non-standard-events), but I guess it needs some work.

I will make the proper changes (including the unit testing), and perhaps file an issue in chisme to improve the docs if that's okay.

Reply:
In your PR you are suggesting to remove the `refresh_event` arg from the `Class`. I think we should keep it to not limit the library's functionality, or use `*args`. wdyt?

Reply:
Updated, wdyt?

Reply:
Yeah sorry it is not really clear how to use this.  I'd also accept the criticism of "To set the events I'm observing, why am I overwriting the private attribute `self._events_to_observe`?" :D  It shouldn't be this way, we just ran out of time before and never came back to make this better.  

Your PR as of now lgtm for this.  I'm guessing you came around on removing the `refresh_event`?  To me, since this code will always get used only for this specific setup, leaving that arg out is ok

Reply:
The updates lgtm in terms of function, but I don't see the value keeping the `refresh_event` arg.  Are there events other than `relation_changed` and `relation_broken` that should trigger `KubernetesServiceInfoRequirer`?  I don't think there are.

Reply:
Re `refresh_event` -  my intention is to keep the component as generic as possible, I see this being abstracted to chisme eventually, so I had it there. You do have a point and perhaps we don't really need it, at least for this charm we don't. I don't have a hard opinion on this one, though.

Reply:
Updated, let me know what you think

Reply:
I think this makes more sense.  But if you do reuse it and have to add `refresh_event` back in, I grant you the right to say "I told you so" :)